### PR TITLE
fix(dynamic-config): use `matches_any_origin` to scrub HTTP hosts in spans

### LINKED ISSUE #3936: `relay.span-normalization.allowed_hosts` global options failed to deserialized
I implemented this a while back: https://github.com/getsentry/relay/pull/3813

On 24.8.0 (self-hosted), it turns out `String` or `&str` can't be properly deserialized into `Host` type. There's this upstream issue on the `url` crate: https://github.com/servo/rust-url/issues/543 and  https://github.com/servo/rust-url/issues/895.

The solution is quite simple, we'll just have to override the `deserialize` method for `Host` specifically. We'll just need to implement a custom deserializer function and change this specific line:
https://github.com/getsentry/relay/blob/92c008e2893f50c2ff9d37308b1e020e0f82b2a4/relay-dynamic-config/src/global.rs#L223

I made a working repro:

Cargo.toml
```toml
[package]
name = "rust-host-vec"
version = "0.1.0"
edition = "2021"

[dependencies]
serde_json = "1"
serde = { version = "1.0.159", features = ["derive", "rc"] }
serde-transcode = "1.1.1"
url = { version = "2", features = ["serde"] }
```

src/main.rs
```rust
use std::fmt;

use serde::{de, Deserialize};
use url::Host;

#[derive(Default, Clone, Debug, Deserialize, PartialEq)]
#[serde(default)]
pub struct GlobalConfig {
    #[serde(
        rename = "relay.span-normalization.allowed_hosts",
        deserialize_with = "deserialize_host",
        skip_serializing_if = "Vec::is_empty"
    )]
    pub http_span_allowed_hosts: Vec<Host>,
}

fn deserialize_host<'de, D>(deserializer: D) -> Result<Vec<Host>, D::Error>
where
    D: serde::de::Deserializer<'de>,
{
    // Stolen from https://users.rust-lang.org/t/need-help-with-serde-deserialize-with/18374
    struct HostVisitor;

    impl<'de> de::Visitor<'de> for HostVisitor {
        type Value = Vec<Host>;

        fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
            formatter.write_str("a string containing host data")
        }

        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
        where
            A: de::SeqAccess<'de>,
        {
            let mut values = Vec::<Host>::new();
            while let Some(value) = seq.next_element()? {
                let this = Host::parse(value);
                let op = de::Error::custom;
                match this {
                    Ok(t) => values.push(t),
                    Err(e) => return Err(op(e)),
                }
            }

            Ok(values)
        }
    }

    deserializer.deserialize_any(HostVisitor)
}

fn main() {
    let input = r###"{
        "relay.span-normalization.allowed_hosts": [
            "vault.internal.example.com"
        ]
      }"###;

    match serde_json::from_str::<GlobalConfig>(input) {
        Ok(config) => {
            println!("{:?}", config);
        }
        Err(e) => {
            println!("{:?}", e);
        }
    }
}
```

Might work on this later this weekend, or the next weekend. I can't work on this right now since I got work to do. Also I guess I need to write few tests here and there for deserializing the new global option I made 😒

### PR DESCRIPTION
**Pull request by aldy505**
Closes https://github.com/getsentry/relay/issues/3936

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. and is gonna need some rights from me in order to utilize my contributions in this here PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

### CODE REVIEW

**Review by iambriccardo — APPROVED** [review]

File: relay-dynamic-config/src/global.rs [hunk 1]
```diff
@@ -372,6 +373,39 @@ where
...
+        where
+            A: de::SeqAccess<'de>,
+        {
+            let mut values = Vec::<Host>::new();
+            while let Some(value) = seq.next_element()? {
+                let this = Host::parse(value);
+                if let Ok(t) = this {
+                    values.push(t)
+                }
+            }
```
Comment (Dav1dde) [inline relay-dynamic-config/src/global.rs]: This visitor still fails if either one of the elements in the list has the wrong type or even the entire object is of the wrong type (more likely).

We can keep this fallible and not ignore invalid hosts (it's not user config, they should be always valid) and instead handle the entire error of the visitor at the `deserialize.deserialize_*` call.

File: relay-dynamic-config/src/global.rs [hunk 2]
```diff
@@ -372,6 +373,39 @@ where
...
+
+            Ok(values)
+        }
+    }
+
+    deserializer.deserialize_any(HostVisitor)
```
Comment (Dav1dde) [inline relay-dynamic-config/src/global.rs]:
```suggestion
    deserializer.deserialize_seq(HostVisitor)
```

Can handle the error here.

**Review by Dav1dde — COMMENTED** [review]
There is also the option we move away from `Host` and use `matches_any_origin` https://github.com/getsentry/relay/blob/5f11c71210b6b9c7f903ac3f1b755a4af1d7af8f/relay-filter/src/csp.rs#L121 for matching and deserialize into `SchemeDomainPort`.

@jjbayer wdyt?

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+**Bug Fixes**:
+
+- Use custom `Vec<Host>` deserialize method for parsing dynamic config ([#3939](https://github.com/getsentry/relay/pull/3939))
```
Comment (jjbayer) [inline CHANGELOG.md]:
```suggestion
- Use `matches_any_origin` to scrub HTTP hosts in spans. ([#3939](https://github.com/getsentry/relay/pull/3939)).
```

**Comment by jjbayer** [comment]
> There is also the option we move away from `Host` and use `matches_any_origin`
> 
> https://github.com/getsentry/relay/blob/5f11c71210b6b9c7f903ac3f1b755a4af1d7af8f/relay-filter/src/csp.rs#L121
> 
> for matching and deserialize into `SchemeDomainPort`.
> @jjbayer wdyt?

Good idea, this would even come with wildcard matching built in. Parsing into `SchemaDomainPort` for every span is not ideal, but at least we know it works, i.e. @aldy505 you could basically copy-paste [what `is_valid_origin` does](https://github.com/getsentry/relay/blob/97606bfa8332e8cdecf465a8e94df893508979ba/relay-server/src/services/project/state/info.rs#L150-L169).

**Comment by aldy505** [comment]
@Dav1dde @jjbayer good morning, can you please re-run the failing CI? it's kind of flaky.

Another note: the `SchemeHostPort` can't do well with IP globs, so `172.16.0.*` is invalid. But I think that should be tackled on another PR.

File: relay-event-normalization/src/normalize/span/description/mod.rs [hunk 4]
```diff
@@ -1232,4 +1238,51 @@ mod tests {
...
+                "http://192.168.1.1:3000",
+                &["192.168.1.1".to_string()],
+                "http://192.168.1.1:3000",
+            ),
+            (
+                "http://[1fff:0:a88:85a3::ac1f]:8001/foo",
+                &["[1fff:0:a88:85a3::ac1f]".to_string()],
+                "http://[1fff:0:a88:85a3::ac1f]:8001",
+            ),
+        ];
```
Comment (jjbayer) [inline relay-event-normalization/src/normalize/span/description/mod.rs]:
```suggestion
        let examples = [
            (
                "https://foo.bar.internal/api/v1/submit",
                ["foo.bar.internal".to_string()],
                "https://foo.bar.internal",
            ),
            (
                "http://192.168.1.1:3000",
                ["192.168.1.1".to_string()],
                "http://192.168.1.1:3000",
            ),
            (
                "http://[1fff:0:a88:85a3::ac1f]:8001/foo",
                ["[1fff:0:a88:85a3::ac1f]".to_string()],
                "http://[1fff:0:a88:85a3::ac1f]:8001",
            ),
        ];
```

This should help get rid of the clippy allow.

File: relay-event-normalization/Cargo.toml [hunk 5]
```diff
@@ -29,6 +29,7 @@ relay-log = { workspace = true }
 relay-protocol = { workspace = true }
 relay-statsd = { workspace = true }
 relay-ua = { workspace = true }
+relay-filter = { workspace = true }
```
Comment (jjbayer) [inline relay-event-normalization/Cargo.toml:32]: nit: We should move the utility function to `relay-common` instead of pulling in `relay-filter` here.

**Review by jjbayer — APPROVED** [review]
Please update the title to match the new behavior.

**Comment by aldy505** [comment]
@jjbayer all greens now

**Comment by jjbayer** [comment]
@aldy505 we had to revert this PR because it caused panics for URLs that contain `[]` brackets in their path. I'm working on a fix now to replace the manual parser with `url::parse`.

**Pull request merged**